src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,229 @@
+
+            if (ErrorHandler.Failed(StartTrackingView(primaryTextView)))
+                Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+
+            if (ErrorHandler.Succeeded(codeWindow.GetSecondaryView(out var secondaryTextView)))
+            {
+                if (ErrorHandler.Failed(StartTrackingView(secondaryTextView)))
+                    Debug.Fail("StartTrackingView failed during DocumentOutlineControl initialization.");
+            }
+
+            StartComputeUIModelTask();
+        }
+
+        int IVsCodeWindowEvents.OnNewView(IVsTextView pView)
+        {
+            ThreadingContext.ThrowIfNotOnUIThread();
+            StartTrackingView(pView);
+            return VSConstants.S_OK;
+        }
+
+        private int StartTrackingView(IVsTextView textView)
+        {
+            ThreadingContext.ThrowIfNotOnUIThread();
+
+            var wpfTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+            if (wpfTextView is null)
+                return VSConstants.E_FAIL;
+
+            _trackedTextViews.Add(textView, wpfTextView);
+
+            wpfTextView.Caret.PositionChanged += Caret_PositionChanged;
+
+            // Subscribe only once since text buffer is the same for the primary and secondary text views.
+            if (_trackedTextViews.Count == 1)
+                wpfTextView.TextBuffer.Changed += TextBuffer_Changed;
+
+            return VSConstants.S_OK;
+        }
+
+        int IVsCodeWindowEvents.OnCloseView(IVsTextView pView)
+        {
+            ThreadingContext.ThrowIfNotOnUIThread();
+
+            if (_trackedTextViews.TryGetValue(pView, out var view))
+            {
+                view.Caret.PositionChanged -= Caret_PositionChanged;
+
+                // Unsubscribe only once since text buffer is the same for the primary and secondary text views.
+                if (_trackedTextViews.Count == 1)
+                    view.TextBuffer.Changed -= TextBuffer_Changed;
+
+                _trackedTextViews.Remove(pView);
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// On text buffer change, obtain an updated UI model and update the view.
+        /// </summary>
+        private void TextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+            => StartComputeUIModelTask();
+
+        /// <summary>
+        /// On caret position change in a text view, highlight the corresponding symbol node in the window.
+        /// </summary>
+        private void Caret_PositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            if (!e.NewPosition.Equals(e.OldPosition))
+                StartHightlightNodeTask();
+        }
+
+        private void ExpandAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolItem>)symbolTree.ItemsSource, true);
+        }
```

Comment:
So I think we need to add a bool property to `DocumentSymbolModel` that tracks `IsCollapseAll`. Otherwise, when a new model comes in from the `_computeUIModelQueue` (aka the developer types something) we will automatically over-ride the expand/collapse settings right? Presumably this should just queue a new item for StartUpdateUIModelTask but pass in a bool for expand/collapse

Reply:
I think this needs to be added as a property to `DocumentOutlineControl.xaml.cs` just like `SortOption`. Otherwise, if we pass in a bool to `_computeUIModelQueue` to collapse everything then change the text buffer, it's unclear whether nodes should be collapsed or expanded the next time we add work to the queue.

Reply:
Wait that doesn't work either... if the user collapses/expands a node and types, all the nodes collapse/expand again.